rasa_core/actions/action.py

```diff
@@ -377,3 +383,113 @@ def __init__(self, action_name, message=None):
 
     def __str__(self):
         return self.message
+
+
+class ActionRevertFallbackEvents(Action):
+    """Reverts events which were done during the `TwoStageFallbackPolicy`.
+
+       This reverts user messages and bot utterances done during a fallback
+       of the `TwoStageFallbackPolicy`. By doing so it is not necessary to
+       write custom stories for the different paths, but only of the happy
+       path.
+    """
+
+    def name(self) -> Text:
+        return ACTION_REVERT_FALLBACK_EVENTS_NAME
+
+    def run(self, dispatcher: 'Dispatcher', tracker: 'DialogueStateTracker',
+            domain: 'Domain') -> List[Event]:
+        from rasa_core.policies.two_stage_fallback import (has_user_clarified,
+                                                           has_user_confirmed)
+        import copy
+
+        last_user_event = tracker.latest_message.intent.get('name')
+        revert_events = []
+
+        # User confirmed
+        if has_user_confirmed(last_user_event, tracker):
+            revert_events = _revert_confirmation_events()
```

Comment:
as a last step, I'd put the code for the three `if` cases into separate functions